crypto/bn/bn_dsa.c

```diff
@@ -0,0 +1,140 @@
+static int app_data_index = 0;
+static size_t get_entropy(RAND_DRBG *drbg, unsigned char **pout,
+                          int entropy, size_t min_len, size_t max_len,
+                          int prediction_resistance)
+{
+    DSA_DRBG_CTX *dsa_drbg_ctx = (DSA_DRBG_CTX *)RAND_DRBG_get_ex_data(drbg, app_data_index);
+
+    *pout = (unsigned char *)dsa_drbg_ctx->x;
+    return dsa_drbg_ctx->rlen;
+}
+static size_t get_nonce(RAND_DRBG *drbg, unsigned char **pout,
+                        int entropy, size_t min_len, size_t max_len)
+{
+    DSA_DRBG_CTX *dsa_drbg_ctx = (DSA_DRBG_CTX *)RAND_DRBG_get_ex_data(drbg, app_data_index);
+
+    *pout = (unsigned char *)dsa_drbg_ctx->h;
+    return dsa_drbg_ctx->rlen;
+}
+
+static int bits2int(BIGNUM *out, int qlen,
+    const unsigned char *message, size_t message_len)
+{
+
+    if (BN_bin2bn(message, (int)message_len, out) == NULL)
+        return 0;
+    if ((int)message_len * 8 > qlen)
+        return BN_rshift(out, out, (int)message_len * 8 - qlen);
+    return 1;
+}
+static int int2octects(unsigned char *out, const BIGNUM  *num, int rlen)
+{
+    return BN_bn2binpad(num, out, rlen);
+}
+static int bits2octects(unsigned char *out, const BIGNUM  *range,
+    const unsigned char *message, size_t message_len, BN_CTX  *ctx)
+{
+    BIGNUM *num = BN_new();
+
+    if (
+        !bits2int(num, BN_num_bits(range), message, message_len) ||
+        !BN_mod(num, num, range, ctx) ||
+        !BN_bn2binpad(num, out, BN_num_bytes(range)))
+    {
+        BN_free(num);
+        return 0;
+    }
+
+    BN_clear_free(num);
+
+    return 1;
+}
+
+int bn_generate_dsa_deterministic_nonce(BIGNUM *out, const BIGNUM *range,
+    const BIGNUM *priv, const unsigned char *message,
+    size_t message_len, int hash_type, BN_CTX *ctx)
+{
+    int ret = 0;
+
+    RAND_DRBG *drbg = NULL;
+    DSA_DRBG_CTX dsa_drbg_ctx;
+    unsigned int drbg_flags = RAND_DRBG_FLAG_HMAC;
+    int rlen = 0, qlen = 0;
+    unsigned char *x = NULL, *h = NULL, *T = NULL;
+
+    app_data_index = RAND_DRBG_get_ex_new_index(CRYPTO_EX_INDEX_DSA, NULL, NULL, NULL, NULL);
+
+    if ((qlen = BN_num_bits(range)) < 0
+        || (rlen = BN_num_bytes(range)) < 0)
+        goto cleanup;
+
+    if ((drbg = RAND_DRBG_new(hash_type, drbg_flags, NULL)) == NULL
+        || !RAND_DRBG_set_callbacks(drbg, get_entropy, NULL, get_nonce, NULL))
+        goto cleanup;
+
+    if ((x = (unsigned char*)OPENSSL_malloc(rlen)) == NULL
+        || (h = (unsigned char*)OPENSSL_malloc(rlen)) == NULL
+        || (T = (unsigned char*)OPENSSL_malloc(rlen)) == NULL)
+        goto cleanup;
+
+    memset(&dsa_drbg_ctx, 0, sizeof(dsa_drbg_ctx));
+
+    if (!int2octects(x, priv, rlen)
+        || !bits2octects(h, range, message, message_len, ctx))
+        goto cleanup;
+
+    dsa_drbg_ctx.x = x;
+    dsa_drbg_ctx.h = h;
+    dsa_drbg_ctx.rlen = rlen;
+
+    /* TODO(3.0): Rewrite the following two lines and include "../rand/rand_lcl.h" if drbg min_lengths can be set using methods */
```

Comment:
Is this todo copied from other code? It probably doesnt apply to this case?

Reply:
@slontis Nope, it is added by me. I think that set min_lengths in this way is ugly and needs to change! However, it is not possible to set min_lenngths using methods, so I added this TODO.